### PR TITLE
Collect information about expenses

### PR DESCRIPTION
**Pull request by jayroh**
Work in progress branch. More to come!

### CODE REVIEW

**Comment by jayroh** [comment]
[As a client, I want to provide information about my expenses to maximize the benefits I will receive](https://trello.com/c/MrLrwn3O/109-as-a-client-i-want-to-provide-information-about-my-expenses-to-maximize-the-benefits-i-will-receive)

File: app/controllers/expenses_additional_controller.rb [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+  def checkboxes_to_array(checkboxes)
+    checkboxes.select { |k| step_params[k].in?(["1", 1, true]) }.map(&:to_s)
+  end
+
+  def skip?
+    @_skip ||= previous_page_all_false?
+  end
+
+  def previous_page_all_false?
+    sections.values.all?(&:!)
```
Comment (jessieay) [inline app/controllers/expenses_additional_controller.rb]: oooh fancy

File: app/controllers/expenses_additional_controller.rb [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+  end
+
+  def sections
+    @sections ||= current_snap_application.
+      attributes.
+      slice(
+        "court_ordered",
```
Comment (jessieay) [inline app/controllers/expenses_additional_controller.rb]: should these go into a constant too?

File: app/controllers/expenses_additional_sources_controller.rb [hunk 3]
```diff
@@ -0,0 +1,13 @@
+# frozen_string_literal: true
+
+class ExpensesAdditionalSourcesController < StandardStepsController
+  private
+
+  def step_params
+    super.tap do |params_hash|
+      params_hash.each do |key, val|
+        params_hash[key] = val == "true" ? true : false
```
Comment (jessieay) [inline app/controllers/expenses_additional_sources_controller.rb]: this might be easier to read in a non-ternary format.

File: app/controllers/expenses_introduction_controller.rb [hunk 4]
```diff
@@ -0,0 +1,4 @@
+# frozen_string_literal: true
+
+class ExpensesIntroductionController < StepsController
```
Comment (jessieay) [inline app/controllers/expenses_introduction_controller.rb]: when I created this controller and the other "static" pages controllers I overrode `update` but I guess I didn't have to! *magic* ✨

File: app/steps/expenses_additional.rb [hunk 5]
```diff
@@ -0,0 +1,24 @@
+# frozen_string_literal: true
+
+class ExpensesAdditional < Step
+  step_attributes \
```
Comment (jessieay) [inline app/steps/expenses_additional.rb]: super tiny stylistic thing: in other step classes we've been putting parens around step_attributes attributes rather than using `\` -- also, alpha order?

File: app/views/expenses_additional/edit.html.erb [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+      <div class="form-card__content">
+        <h4 class="step-section-header__headline">
+          Care expenses
+        </h4>
+
+        <%= f.mb_input_field :monthly_care_expenses,
+          "In total, how much do you pay in care expenses each month?" %>
+
+        <%= f.mb_checkbox_set [
+          { method: :childcare,           label: 'Childcare'           },
```
Comment (jessieay) [inline app/views/expenses_additional/edit.html.erb]: not sure I'm on board with this type of spacing for hashes, but I could be convinced. 💟

File: app/views/expenses_additional/edit.html.erb [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+          { method: :elderly_care,        label: 'Elderly Care'        },
+          { method: :disabled_adult_care, label: 'Disabled adult care' },
+          ], label_text: 'What kind of care expenses do you have?' %>
```
Comment (jessieay) [inline app/views/expenses_additional/edit.html.erb]: seems like there's some inconsistency in single/double quotes in this file. Think we're doing double?

File: db/migrate/20170816143355_add_expense_columns_to_snap_application.rb [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+    add_column :snap_applications, :utility_heat, :boolean
+    add_column :snap_applications, :utility_cooling, :boolean
+    add_column :snap_applications, :utility_electrity, :boolean
+    add_column :snap_applications, :utility_water_sewer, :boolean
+    add_column :snap_applications, :utility_trash, :boolean
+    add_column :snap_applications, :utility_phone, :boolean
+    add_column :snap_applications, :utility_other, :boolean
+
+    add_column :snap_applications, :dependent_care, :boolean
+    add_column :snap_applications, :medical, :boolean
```
Comment (jessieay) [inline db/migrate/20170816143355_add_expense_columns_to_snap_application.rb]: I feel like a bool field called `medical` is not super clear -- thoughts on a more descriptive name for this column? (feel similarly about `dependent_card` `court_ordered` and `tax_deductible`)

File: app/steps/expenses_additional_source.rb [hunk 9]
```diff
@@ -0,0 +1,19 @@
+# frozen_string_literal: true
+
+class ExpensesAdditionalSource < Step
+  step_attributes \
+    :dependent_care,
+    :medical,
+    :court_ordered,
+    :tax_deductible
+
+  validates \
```
Comment (jessieay) [inline app/steps/expenses_additional_source.rb]: should we have a controller spec to test this validation?

File: app/controllers/expenses_additional_controller.rb [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+    tax_deductible_expenses: %w[
+      student_loan_interest
+      other_tax_deductible
+    ],
+  }.freeze
+
+  def edit
+    set_sections
+
+    attribute_array = ARRAY_ATTRIBUTES.keys.flat_map do |key|
```
Comment (jessieay) [inline app/controllers/expenses_additional_controller.rb]: maybe this should happen in a well-named private method? there is enough happening in these controller methods that I am getting a but turned around while reading them

File: app/controllers/expenses_additional_sources_controller.rb [hunk 11]
```diff
@@ -0,0 +1,13 @@
+# frozen_string_literal: true
+
+class ExpensesAdditionalSourcesController < StandardStepsController
+  private
+
```
Comment (jessieay) [inline app/controllers/expenses_additional_sources_controller.rb]: we've been defining the `existing_attributes` method to ensure that attrs are filled in on the page when someone returns. Is that functionality working here? Perhaps a controller spec would be a good way to confirm

File: db/migrate/20170816143355_add_expense_columns_to_snap_application.rb (hunk 8 above)
Comment (jessieay) [inline db/migrate/20170816143355_add_expense_columns_to_snap_application.rb]: even `medical_expenses` as a boolean would be clearer! not sure we need `has`

File: app/controllers/expenses_additional_sources_controller.rb (hunk 11 above)
Comment (jayroh) [inline app/controllers/expenses_additional_sources_controller.rb]: @jessieay what's the rule for the `existing_attributes` method? Is the goal to provide a hash with the existing model's current state as it pertains to *this* action?  So for here I would be adding?

```ruby
  def existing_attributes
    HashWithIndifferentAccess.new(application_attributes)
  end

  def application_attributes
    app = current_snap_application

    {
      monthly_care_expenses: app.monthly_care_expenses,
      monthly_medical_expenses: app.monthly_medical_expenses,
      monthly_court_ordered_expenses: app.monthly_court_ordered_expenses,
      monthly_tax_deductible_expenses: app.monthly_tax_deductible_expenses,
      care_expenses: app.care_expenses,
      medical_expenses: app.medical_expenses,
      court_ordered_expenses: app.court_ordered_expenses,
      tax_deductible_expenses: app.tax_deductible_expenses,
    }
  end
```

Is that right?  And is that tested with that shared example?

File: app/steps/expenses_additional_source.rb (hunk 9 above)
Comment (jessieay) [inline app/steps/expenses_additional_source.rb]: yeah unit test works, too! just wanted to make sure we have *something* testing this validation

File: app/controllers/expenses_additional_sources_controller.rb (hunk 11 above)
Comment (jessieay) [inline app/controllers/expenses_additional_sources_controller.rb]: yeah, we've been testing this at the controller level so far. Example: https://github.com/codeforamerica/michigan-benefits/blob/13c20027a88425746d695df75abee0040a158da2/spec/controllers/household_add_member_controller_spec.rb#L14

File: db/migrate/20170816143355_add_expense_columns_to_snap_application.rb (hunk 8 above)
Comment (jayroh) [inline db/migrate/20170816143355_add_expense_columns_to_snap_application.rb]: There's already a `medical_expenses` column (an array of the actual medical expenses they're incurring) versus `medical` (the bool)

File: app/views/expenses_introduction/edit.html.erb [hunk 12]
```diff
@@ -1,36 +1,35 @@
...
-<div class='form-card'>
-  <header class='form-card__header'>
-    <div class='step-section-header'>
-      <div class="step-section-header__icon illustration illustration--section-3"></div>
-      <div class='step-section-header__headline'>You're halfway there!</div>
-      <div class='step-section-header__subhead'>
+<div class="form-card">
+  <header class="form-card__header">
+    <div class="step-section-header">
+      <% icon 'section-3' %>
```
Comment (jessieay) [inline app/views/expenses_introduction/edit.html.erb]: this `icon` helper does not exist in this codebase, needs to be `<div class="step-section-header__icon illustration illustration--section-3"></div>` per diff

File: db/migrate/20170816143355_add_expense_columns_to_snap_application.rb (hunk 8 above)
Comment (jessieay) [inline db/migrate/20170816143355_add_expense_columns_to_snap_application.rb]: hmmm ok, let's potentially revisit this soon, then. I am just noticing as I fill in the PDF that it is hard to sometimes tell which DB fields match to which PDF fields. Hopefully we can get to a place naming-wise where it is more obvious without having to dig through the views and controllers.

File: spec/controllers/expenses_additional_controller_spec.rb [hunk 13]
```diff
@@ -0,0 +1,42 @@
+# frozen_string_literal: true
+
+require "rails_helper"
+
+RSpec.describe ExpensesAdditionalController, type: :controller do
+  let(:step) { assigns(:step) }
+  let(:step_class) { ExpensesAdditional }
+  before {}
```
Comment (jessieay) [inline spec/controllers/expenses_additional_controller_spec.rb]: surprised we don't need a `current_snap_apploication` set for this page to render?

Comment (jessieay) [inline spec/controllers/expenses_additional_controller_spec.rb]: oh I see -- we are setting in each spec. In other controller specs we've been doing this via a `before` action. I generally don't love `before` but it kinda makes sense here since they all need it

File: spec/steps/expenses_additional_source_spec.rb [hunk 14]
```diff
@@ -0,0 +1,32 @@
...
+    )
+
+    valid_falses_step = ExpensesAdditionalSource.new(
+      court_ordered: false,
+      dependent_care: false,
+      medical: false,
+      tax_deductible: false,
+    )
+
+    invalid_step = ExpensesAdditionalSource.new(
```
Comment (jessieay) [inline spec/steps/expenses_additional_source_spec.rb]: could also test for `nil` values if you thought that was valuable

File: spec/controllers/expenses_additional_controller_spec.rb (hunk 13 above)
Comment (jayroh) [inline spec/controllers/expenses_additional_controller_spec.rb]: d'oh. Shoulda removed that 😊
